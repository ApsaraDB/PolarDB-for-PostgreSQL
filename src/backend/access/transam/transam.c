/*-------------------------------------------------------------------------
 *
 * transam.c
 *	  postgres transaction (commit) log interface routines
 *
 * Support CTS-based transactions.
 * Author: Junbin Kang
 *
 * Portions Copyright (c) 2020, Alibaba Group Holding Limited
 * Portions Copyright (c) 1996-2018, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 *
 * IDENTIFICATION
 *	  src/backend/access/transam/transam.c
 *
 *-------------------------------------------------------------------------
 */

#include "postgres.h"

#include "access/clog.h"
#include "access/ctslog.h"
#include "access/mvccvars.h"
#include "access/subtrans.h"
#include "access/transam.h"
#include "distributed_txn/txn_timestamp.h"
#include "storage/lmgr.h"
#include "utils/snapmgr.h"

/*
 * Single-item cache for results of TransactionIdGetCommitSeqNo.  It's worth
 * having
 * such a cache because we frequently find ourselves repeatedly checking the
 * same XID, for example when scanning a table just after a bulk insert,
 * update, or delete.
 */
static TransactionId cachedFetchXid = InvalidTransactionId;
static CommitSeqNo cachedCSN;

/*
 * Also have a (separate) cache for CLogGetCommitLSN()
 */
static TransactionId cachedLSNFetchXid = InvalidTransactionId;
static XLogRecPtr cachedCommitLSN;

/*
 * TransactionIdGetCommitSeqNo --- fetch CSN of specified transaction id
 */
CommitSeqNo
TransactionIdGetCommitSeqNo(TransactionId transactionId)
{
	CommitSeqNo csn;

	/*
	 * Before going to the commit log manager, check our single item cache to
	 * see if we didn't just check the transaction status a moment ago.
	 */
	if (TransactionIdEquals(transactionId, cachedFetchXid))
		return cachedCSN;

	/*
	 * Also, check to see if the transaction ID is a permanent one.
	 */
	if (!TransactionIdIsNormal(transactionId))
	{
		if (TransactionIdEquals(transactionId, BootstrapTransactionId))
			return COMMITSEQNO_FROZEN;
		if (TransactionIdEquals(transactionId, FrozenTransactionId))
			return COMMITSEQNO_FROZEN;
		return COMMITSEQNO_ABORTED;
	}

	/*
	 * If the XID is older than TransactionXmin, check the clog. Otherwise
	 * check the csnlog.
	 */
#ifndef ENABLE_DISTRIBUTED_TRANSACTION
	Assert(TransactionIdIsValid(TransactionXmin));
	if (TransactionIdPrecedes(transactionId, TransactionXmin))
	{
		XLogRecPtr	lsn;

		if (CLogGetStatus(transactionId, &lsn) == CLOG_XID_STATUS_COMMITTED)
			csn = COMMITSEQNO_FROZEN;
		else
			csn = COMMITSEQNO_ABORTED;
	}
	else
#endif
	{
		csn = CTSLogGetCommitTs(transactionId);

		if (csn == COMMITSEQNO_COMMITTING)
		{
			/*
			 * If the transaction is committing at this very instant, and
			 * hasn't set its CSN yet, wait for it to finish doing so.
			 *
			 * XXX: Alternatively, we could wait on the heavy-weight lock on
			 * the XID. that'd make TransactionIdCommitTree() slightly
			 * cheaper, as it wouldn't need to acquire CommitSeqNoLock (even
			 * in shared mode).
			 */

#ifdef ENABLE_DISTRIBUTED_TRANSACTION
			/*
			 * We choose to wait for the specific transaction to run to
			 * completion, so as the transaction commit path does not need to
			 * acquire CommitSeqNoLock.
			 */
#ifdef ENABLE_DISTR_DEBUG
			if (enable_timestamp_debug_print)
				elog(LOG, "wait for committing transaction xid %d to complete", transactionId);
#endif
			XactLockTableWait(transactionId, NULL, NULL, XLTW_None);
#else
			LWLockAcquire(CommitSeqNoLock, LW_EXCLUSIVE);
			LWLockRelease(CommitSeqNoLock);
#endif

			csn = CTSLogGetCommitTs(transactionId);
			Assert(csn != COMMITSEQNO_COMMITTING);
		}
	}

	/*
	 * Cache it, but DO NOT cache status for unfinished transactions! We only
	 * cache status that is guaranteed not to change.
	 */
	if (COMMITSEQNO_IS_COMMITTED(csn) ||
		COMMITSEQNO_IS_ABORTED(csn))
	{
		cachedFetchXid = transactionId;
		cachedCSN = csn;
	}

	return csn;
}

/*
 * TransactionIdDidCommit
 *		True iff transaction associated with the identifier did commit.
 *
 * Note:
 *		Assumes transaction identifier is valid and exists in clog.
 */
bool							/* true if given transaction committed */
TransactionIdDidCommit(TransactionId transactionId)
{
	CommitSeqNo csn;

	csn = TransactionIdGetCommitSeqNo(transactionId);

#ifdef ENABLE_DISTR_DEBUG
	/* perform check against clog for debugging purpose */
	{
		if (COMMITSEQNO_IS_COMMITTED(csn))
		{
			XLogRecPtr	result;
			CLogXidStatus clogstatus;

			clogstatus = CLogGetStatus(transactionId, &result);

			if (clogstatus != CLOG_XID_STATUS_COMMITTED)
				elog(PANIC, "cts log status " UINT64_FORMAT " does not match clog %d",
					 csn, clogstatus);
		}
		else if (COMMITSEQNO_IS_ABORTED(csn))
		{
			XLogRecPtr	result;
			CLogXidStatus clogstatus;

			clogstatus = CLogGetStatus(transactionId, &result);

			if (clogstatus != CLOG_XID_STATUS_ABORTED)
				elog(PANIC, "cts log status " UINT64_FORMAT " does not match clog %d",
					 csn, clogstatus);
		}
	}
#endif

	if (COMMITSEQNO_IS_COMMITTED(csn))
		return true;
	else
		return false;
}

/*
 * TransactionIdDidAbort
 *		True iff transaction associated with the identifier did abort.
 *
 * Note:
 *		Assumes transaction identifier is valid and exists in clog.
 */
bool							/* true if given transaction aborted */
TransactionIdDidAbort(TransactionId transactionId)
{
	CommitSeqNo csn;

	csn = TransactionIdGetCommitSeqNo(transactionId);

#ifdef ENABLE_DISTR_DEBUG
	/* perform check against clog for debugging purpose */
	{
		if (COMMITSEQNO_IS_ABORTED(csn))
		{
			XLogRecPtr	result;
			CLogXidStatus clogstatus;

			clogstatus = CLogGetStatus(transactionId, &result);

			if (clogstatus != CLOG_XID_STATUS_ABORTED)
				elog(PANIC, "cts log status " UINT64_FORMAT " does not match clog %d",
					 csn, clogstatus);
		}
		else if (COMMITSEQNO_IS_COMMITTED(csn))
		{
			XLogRecPtr	result;
			CLogXidStatus clogstatus;

			clogstatus = CLogGetStatus(transactionId, &result);

			if (clogstatus != CLOG_XID_STATUS_COMMITTED)
				elog(PANIC, "cts log status " UINT64_FORMAT " does not match clog %d",
					 csn, clogstatus);
		}
	}
#endif

	if (COMMITSEQNO_IS_ABORTED(csn))
		return true;
	else
		return false;
}

/*
 * Returns the status of the tranaction.
 *
 * Note that this treats a a crashed transaction as still in-progress,
 * until it falls off the xmin horizon.
 */
TransactionIdStatus
TransactionIdGetStatus(TransactionId xid)
{
	CommitSeqNo csn;
	TransactionIdStatus status;

	csn = TransactionIdGetCommitSeqNo(xid);

	if (COMMITSEQNO_IS_COMMITTED(csn))
		status = XID_COMMITTED;
	else if (COMMITSEQNO_IS_ABORTED(csn))
		status = XID_ABORTED;
	else
		status = XID_INPROGRESS;

#ifdef ENABLE_DISTR_DEBUG
	/* perform check against clog for debugging purpose */
	{
		XLogRecPtr	result;
		CLogXidStatus clogstatus;

		clogstatus = CLogGetStatus(xid, &result);

		switch (status)
		{
			case XID_COMMITTED:
				if (clogstatus != CLOG_XID_STATUS_COMMITTED)
					elog(PANIC, "cts log status %d does not match clog %d",
						 status, clogstatus);
				break;
			case XID_ABORTED:
				if (clogstatus != CLOG_XID_STATUS_ABORTED)
					elog(PANIC, "cts log status %d does not match clog %d",
						 status, clogstatus);
				break;
		}
	}
#endif

	return status;
}

/*
 * TransactionIdCommitTree
 *		Marks the given transaction and children as committed
 *
 * "xid" is a toplevel transaction commit, and the xids array contains its
 * committed subtransactions.
 */
void
TransactionIdCommitTree(TransactionId xid, int nxids, TransactionId *xids)
{
	TransactionIdAsyncCommitTree(xid, nxids, xids, InvalidXLogRecPtr);
}

/*
 * TransactionIdAsyncCommitTree
 *		Same as above, but for async commits.
 *
 * "xid" is a toplevel transaction commit, and the xids array contains its
 * committed subtransactions.
 */
void
TransactionIdAsyncCommitTree(TransactionId xid, int nxids, TransactionId *xids,
							 XLogRecPtr lsn)
{
#ifndef ENABLE_DISTRIBUTED_TRANSACTION
	CommitSeqNo csn;
#endif
	TransactionId latestXid;
	TransactionId currentLatestCompletedXid;

	latestXid = TransactionIdLatest(xid, nxids, xids);

	/*
	 * First update the clog, then CSN log. oldestActiveXid advances based on
	 * CSN log content (see AdvanceOldestActiveXid), and it should not become
	 * greater than our xid before we set the clog status. Otherwise other
	 * transactions could see us as aborted for some time after we have
	 * written to CSN log, and somebody advanced the oldest active xid past
	 * our xid, but before we write to clog.
	 */
#ifdef ENABLE_DISTR_DEBUG
	CLogSetTreeStatus(xid, nxids, xids,
					  CLOG_XID_STATUS_COMMITTED,
					  lsn);
#endif

	/*
	 * Grab the CommitSeqNoLock, in shared mode. This is only used to provide
	 * a way for a concurrent transaction to wait for us to complete (see
	 * TransactionIdGetCommitSeqNo()).
	 *
	 * XXX: We could reduce the time the lock is held, by only setting the CSN
	 * on the top-XID while holding the lock, and updating the sub-XIDs later.
	 * But it doesn't matter much, because we're only holding it in shared
	 * mode, and it's rare for it to be acquired in exclusive mode.
	 */
	LWLockAcquire(CommitSeqNoLock, LW_SHARED);

	/*
	 * First update latestCompletedXid to cover this xid. We do this before
	 * assigning a CSN, so that if someone acquires a new snapshot at the same
	 * time, the xmax it computes is sure to cover our XID.
	 */
	currentLatestCompletedXid = pg_atomic_read_u32(&ShmemVariableCache->latestCompletedXid);
	while (TransactionIdFollows(latestXid, currentLatestCompletedXid))
	{
		if (pg_atomic_compare_exchange_u32(&ShmemVariableCache->latestCompletedXid,
										   &currentLatestCompletedXid,
										   latestXid))
			break;
	}
#ifndef ENABLE_DISTRIBUTED_TRANSACTION

	/*
	 * Mark our top transaction id as commit-in-progress.
	 */
	CSNLogSetCommitSeqNo(xid, 0, NULL, lsn, true, COMMITSEQNO_COMMITTING);

	/* Get our CSN and increment */
	csn = pg_atomic_fetch_add_u64(&ShmemVariableCache->nextCommitSeqNo, 1);
	Assert(csn >= COMMITSEQNO_FIRST_NORMAL);

	/* Stamp this XID (and sub-XIDs) with the CSN */
	CSNLogSetCommitSeqNo(xid, nxids, xids, lsn, true, csn);

#endif
	LWLockRelease(CommitSeqNoLock);
}

/*
 * TransactionIdAbortTree
 *		Marks the given transaction and children as aborted.
 *
 * "xid" is a toplevel transaction commit, and the xids array contains its
 * committed subtransactions.
 *
 * We don't need to worry about the non-atomic behavior, since any onlookers
 * will consider all the xacts as not-yet-committed anyway.
 */
void
TransactionIdAbortTree(TransactionId xid, int nxids, TransactionId *xids)
{
	TransactionId latestXid;
	TransactionId currentLatestCompletedXid;

	latestXid = TransactionIdLatest(xid, nxids, xids);

	currentLatestCompletedXid = pg_atomic_read_u32(&ShmemVariableCache->latestCompletedXid);
	while (TransactionIdFollows(latestXid, currentLatestCompletedXid))
	{
		if (pg_atomic_compare_exchange_u32(&ShmemVariableCache->latestCompletedXid,
										   &currentLatestCompletedXid,
										   latestXid))
			break;
	}
	if (enable_timestamp_debug_print)
		elog(LOG, "abort transaction xid %d", xid);

#ifdef ENABLE_DISTR_DEBUG
	CLogSetTreeStatus(xid, nxids, xids,
					  CLOG_XID_STATUS_ABORTED, InvalidXLogRecPtr);
#endif

	CTSLogSetCommitTs(xid, nxids, xids, InvalidXLogRecPtr, false, COMMITSEQNO_ABORTED);
}

/*
 * TransactionIdPrecedes --- is id1 logically < id2?
 */
bool
TransactionIdPrecedes(TransactionId id1, TransactionId id2)
{
	/*
	 * If either ID is a permanent XID then we can just do unsigned
	 * comparison.  If both are normal, do a modulo-2^32 comparison.
	 */
	int32		diff;

	if (!TransactionIdIsNormal(id1) || !TransactionIdIsNormal(id2))
		return (id1 < id2);

	diff = (int32) (id1 - id2);
	return (diff < 0);
}

/*
 * TransactionIdPrecedesOrEquals --- is id1 logically <= id2?
 */
bool
TransactionIdPrecedesOrEquals(TransactionId id1, TransactionId id2)
{
	int32		diff;

	if (!TransactionIdIsNormal(id1) || !TransactionIdIsNormal(id2))
		return (id1 <= id2);

	diff = (int32) (id1 - id2);
	return (diff <= 0);
}

/*
 * TransactionIdFollows --- is id1 logically > id2?
 */
bool
TransactionIdFollows(TransactionId id1, TransactionId id2)
{
	int32		diff;

	if (!TransactionIdIsNormal(id1) || !TransactionIdIsNormal(id2))
		return (id1 > id2);

	diff = (int32) (id1 - id2);
	return (diff > 0);
}

/*
 * TransactionIdFollowsOrEquals --- is id1 logically >= id2?
 */
bool
TransactionIdFollowsOrEquals(TransactionId id1, TransactionId id2)
{
	int32		diff;

	if (!TransactionIdIsNormal(id1) || !TransactionIdIsNormal(id2))
		return (id1 >= id2);

	diff = (int32) (id1 - id2);
	return (diff >= 0);
}


/*
 * TransactionIdLatest --- get latest XID among a main xact and its children
 */
TransactionId
TransactionIdLatest(TransactionId mainxid,
					int nxids, const TransactionId *xids)
{
	TransactionId result;

	/*
	 * In practice it is highly likely that the xids[] array is sorted, and so
	 * we could save some cycles by just taking the last child XID, but this
	 * probably isn't so performance-critical that it's worth depending on
	 * that assumption.  But just to show we're not totally stupid, scan the
	 * array back-to-front to avoid useless assignments.
	 */
	result = mainxid;
	while (--nxids >= 0)
	{
		if (TransactionIdPrecedes(result, xids[nxids]))
			result = xids[nxids];
	}
	return result;
}


/*
 * TransactionIdGetCommitLSN
 *
 * This function returns an LSN that is late enough to be able
 * to guarantee that if we flush up to the LSN returned then we
 * will have flushed the transaction's commit record to disk.
 *
 * The result is not necessarily the exact LSN of the transaction's
 * commit record!  For example, for long-past transactions (those whose
 * clog pages already migrated to disk), we'll return InvalidXLogRecPtr.
 * Also, because we group transactions on the same clog page to conserve
 * storage, we might return the LSN of a later transaction that falls into
 * the same group.
 */
XLogRecPtr
TransactionIdGetCommitLSN(TransactionId xid)
{
	XLogRecPtr	result;

	/*
	 * Currently, all uses of this function are for xids that were just
	 * reported to be committed by TransactionLogFetch, so we expect that
	 * checking TransactionLogFetch's cache will usually succeed and avoid an
	 * extra trip to shared memory.
	 */
	if (TransactionIdEquals(xid, cachedLSNFetchXid))
		return cachedCommitLSN;

	/* Special XIDs are always known committed */
	if (!TransactionIdIsNormal(xid))
		return InvalidXLogRecPtr;

	/*
	 * Get the transaction status.
	 */
#ifdef ENABLE_DISTR_DEBUG
	result = CLogGetLSN(xid);
#endif

	result = CTSLogGetLSN(xid);

	cachedLSNFetchXid = xid;
	cachedCommitLSN = result;

	return result;
}
