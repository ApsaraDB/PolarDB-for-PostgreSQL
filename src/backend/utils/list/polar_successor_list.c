/*----------------------------------------------------------------------------------------
 *
 * polar_successor_list.c
 *  This list is used to manage free resources.
 *  The resources are allocated as array items,
 *  get the free item from polar_successor_list_pop
 *  and call polar_successor_list_push when the item is released.
 *
 * Copyright (c) 2020, Alibaba Group Holding Limited
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * src/backend/utils/list/polar_successor_list.c
 * ---------------------------------------------------------------------------------------
 */

#include "postgres.h"
#include "utils/polar_successor_list.h"

polar_successor_list *
polar_successor_list_init(void *ptr, int total_items)
{
	int i;
	polar_successor_list *list = (polar_successor_list *)ptr;

	for (i = 0; i < total_items; i++)
		list->items[i] = i + 1;

	list->items[total_items - 1] = POLAR_SUCCESSOR_LIST_NIL;

	list->free_head = 0;

	return list;
}

int
polar_successor_list_pop(polar_successor_list *list)
{
	int free_item = list->free_head;

	if (free_item != POLAR_SUCCESSOR_LIST_NIL)
	{
		list->free_head = list->items[free_item];
		list->items[free_item] = POLAR_SUCCESSOR_LIST_NIL;
	}

	return free_item;
}

void
polar_successor_list_push(polar_successor_list *list, int free_item)
{
	if (list->free_head != POLAR_SUCCESSOR_LIST_NIL)
		list->items[free_item] = list->free_head;
	else
		list->items[free_item] = POLAR_SUCCESSOR_LIST_NIL;

	list->free_head = free_item;
}
