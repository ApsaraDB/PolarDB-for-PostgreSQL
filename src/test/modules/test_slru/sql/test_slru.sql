CREATE EXTENSION test_slru;

-- See README for explanation of arguments:
SELECT test_slru();

SELECT test_slru_slot_size_config();
SELECT test_slru_hash_index(10, 256, 10000);
