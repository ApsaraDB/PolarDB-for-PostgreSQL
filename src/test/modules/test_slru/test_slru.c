#include "postgres.h"

#include <time.h>
#include <stdlib.h>
#include <unistd.h>

#include "utils/guc.h"
#include "access/slru.h"
#include "access/multixact.h"
#include "access/subtrans.h"
#include "access/transam.h"
#include "access/commit_ts.h"
#include "commands/async.h"
#include "fmgr.h"
#include "miscadmin.h"
#include "pgstat.h"
#include "storage/polar_fd.h"
#include "storage/predicate.h"
#include "utils/elog.h"
#include "utils/guc.h"
#include "utils/memutils.h"
#include "utils/polar_local_cache.h"
#include "utils/polar_successor_list.h"

static SlruCtlData test_slru_ctl;
static SlruCtlData slru_hash_index_ctl;

static LWLockMinimallyPadded *test_lock;

FILE * fp;

PG_MODULE_MAGIC;

static void
test_rethrow_error(ErrorData *data)
{
	if (data->elevel >= ERROR)
		PG_RE_THROW();
}

static void
test_successor_list(int total_items)
{
	char *buf = malloc(POLAR_SUCCESSOR_LIST_SIZE(total_items));
	polar_successor_list *list;
	int i, j, k;

	list = polar_successor_list_init((void *)buf, total_items);

	Assert(!POLAR_SUCCESSOR_LIST_EMPTY(list));

	for (i=0; i<total_items; i++)
	{
		j = polar_successor_list_pop(list);

		Assert(i==j);
	}

	Assert(polar_successor_list_pop(list) == POLAR_SUCCESSOR_LIST_NIL);
	Assert(POLAR_SUCCESSOR_LIST_EMPTY(list));

	for (k=0; k<10; k++)
	{
		for (i=0; i<total_items; i++)
		{
			polar_successor_list_push(list, i);
		}
		Assert(!POLAR_SUCCESSOR_LIST_EMPTY(list));

		for (i=total_items; i>0; i--)
		{
			j = polar_successor_list_pop(list);

			Assert(j == i - 1);
		}

		Assert(polar_successor_list_pop(list) == POLAR_SUCCESSOR_LIST_NIL);
		Assert(POLAR_SUCCESSOR_LIST_EMPTY(list));
	}

	free(buf);	
}

static bool
test_slru_page_precedes(int page1, int page2)
{
	return page1 < page2;
}

static void
test_slru_page_physical_exists()
{
	int i;
	int slotno;

	for (i=0; i<64; i++)
	{
		LWLockAcquire(&test_lock->lock, LW_EXCLUSIVE);
		Assert(!polar_slru_page_physical_exists(&test_slru_ctl, i));
		slotno = SimpleLruZeroPage(&test_slru_ctl, i);
		SimpleLruWritePage(&test_slru_ctl, slotno);
		Assert(!test_slru_ctl.shared->page_dirty[slotno]);
		slotno = SimpleLruReadPage(&test_slru_ctl, i, false, InvalidTransactionId);
		
		if (i > 0)
			Assert(polar_slru_page_physical_exists(&test_slru_ctl, i-1));
		LWLockRelease(&test_lock->lock);
	}
}

static void
slru_hash_index_precheck(int slot_num)
{
	SlruShared shared = slru_hash_index_ctl.shared;

	Assert(!POLAR_SUCCESSOR_LIST_EMPTY(shared->polar_free_list));
	Assert(shared->victim_pivot == 0);
}

static void
slru_hash_index_postcheck(int page_num)
{
	int i = 0;
	SlruShared shared = slru_hash_index_ctl.shared;

	/* check from mapping slot_no -> page_no */
	for (i = 0; i < page_num; i++)
	{
		polar_slru_hash_entry *entry;
		int slotno;
	   
		LWLockAcquire(&test_lock->lock, LW_SHARED);	
		slotno	= SimpleLruReadPage(&slru_hash_index_ctl,
					i, true, InvalidTransactionId); 
		entry = hash_search(shared->polar_hash_index, (void *)&i, HASH_FIND, NULL);
		Assert(entry != NULL);
		Assert(entry->pageno == i);
		Assert(entry->slotno == slotno);
		LWLockRelease(&test_lock->lock);
	}
}

static void
test_slru_hash_index_internal(int slot_num, int page_num, int test_num)
{
	int i = 0;
	int slotno = 0;
	SlruShared shared = slru_hash_index_ctl.shared;

	slru_hash_index_precheck(slot_num);
	/* prepare data */
	for (i = 0; i < page_num; i++)
	{
		polar_slru_hash_entry *entry;

		LWLockAcquire(&test_lock->lock, LW_EXCLUSIVE);	
		if (i < slot_num)
			Assert(!POLAR_SUCCESSOR_LIST_EMPTY(shared->polar_free_list));
		else
			Assert(POLAR_SUCCESSOR_LIST_EMPTY(shared->polar_free_list));

		slotno = SimpleLruZeroPage(&slru_hash_index_ctl, i);
		SimpleLruWritePage(&slru_hash_index_ctl, slotno);
		entry = hash_search(shared->polar_hash_index, (void *)&i, HASH_FIND, NULL);
		Assert(entry != NULL);
		Assert(entry->pageno == i);
		Assert(entry->slotno == slotno);
		LWLockRelease(&test_lock->lock);
	}

	slru_hash_index_postcheck(page_num);

	/* run chaos test */
	for (i = 0; i < test_num; i++)
	{
		int r = rand();
		int page = rand() % page_num;
		int invalid_page = 0;
		switch(r % 10)
		{
			case 0:
				/* Zero Page */
				LWLockAcquire(&test_lock->lock, LW_EXCLUSIVE);	
				slotno = SimpleLruZeroPage(&slru_hash_index_ctl, page);
				SimpleLruWritePage(&slru_hash_index_ctl, slotno);
				LWLockRelease(&test_lock->lock);
				break;
			case 1:
			case 2:
				/* ReadPage */
				LWLockAcquire(&test_lock->lock, LW_SHARED);	
				SimpleLruReadPage(&slru_hash_index_ctl, page, true, InvalidTransactionId);
				LWLockRelease(&test_lock->lock);
				break;
			case 3:
				LWLockAcquire(&test_lock->lock, LW_EXCLUSIVE);
				SimpleLruWritePage(&slru_hash_index_ctl, rand() % slot_num);
				LWLockRelease(&test_lock->lock);
				break;
			case 4:
			case 5:
				/* ReadOnly */
				SimpleLruReadPage_ReadOnly(&slru_hash_index_ctl, page, InvalidTransactionId);
				LWLockRelease(&test_lock->lock);
				break;
			case 6:
			case 7:
				/* invalid page */
				invalid_page = shared->page_number[rand() % slot_num];
				polar_slru_invalid_page(&slru_hash_index_ctl, invalid_page);
				break;
			case 8:
				// SimpleLruTruncate(&slru_hash_index_ctl, page);
				// if (page > truncate_page)
				// 	truncate_page = page;
				break;
			case 9:
				SimpleLruFlush(&slru_hash_index_ctl, false);
				break;
		}		
		slru_hash_index_postcheck(page_num);
	}

	slru_hash_index_postcheck(page_num);
}

static void
test_promote_slru_with_cache_err(void)
{
	int i = 0;
	MemoryContext   ccxt = CurrentMemoryContext;

	emit_log_hook = test_rethrow_error;

	PG_TRY();
	{
		polar_slru_promote(&test_slru_ctl);
	}
	PG_CATCH();
	{
		ErrorData *err;
		MemoryContextSwitchTo(ccxt);
		err = CopyErrorData();
		Assert(err->elevel == FATAL);
		pfree(err);
		MemoryContextReset(ccxt);
		i++;
	}
	PG_END_TRY();

	emit_log_hook = NULL;

	Assert(i == 1);
}

static void
test_promote_slru_with_cache_dir_err(void)
{
	polar_local_cache cache;
	uint32 io_permission = POLAR_CACHE_LOCAL_FILE_READ | POLAR_CACHE_LOCAL_FILE_WRITE;
	FILE *fp;
	uint64 seg = 100;

	IsUnderPostmaster = false;
	cache = polar_create_local_cache("test_slru", "pg_test_slru",
			   8, (SLRU_PAGES_PER_SEGMENT * BLCKSZ), LWTRANCHE_POLAR_CSNLOG_LOCAL_CACHE,
			   io_permission, NULL);
	if (cache)
		polar_local_cache_move_trash(cache->dir_name);
	IsUnderPostmaster = true;

	polar_slru_reg_local_cache(&test_slru_ctl, cache);
	fp = fopen("pg_test_slru/readonly_flushed_seg_0001", "w");
	fwrite(&seg, sizeof(uint64), 1, fp);
	fclose(fp);
	/* POLAR: filename '0064' corresponds to segno which is 100. */
	fp = fopen("pg_test_slru/0064", "w");
	fwrite(&seg, sizeof(uint64), 1, fp);
	fclose(fp);
	/* POLAR: remove write io permission of file. */
	chmod("pg_test_slru/0064", S_IRUSR|S_IRGRP|S_IROTH);

	test_promote_slru_with_cache_err();
}

PG_FUNCTION_INFO_V1(test_slru);
/*
 * SQL-callable entry point to perform all tests.
 *
 * If a 1% false positive threshold is not met, emits WARNINGs.
 *
 * See README for details of arguments.
 */
Datum
test_slru(PG_FUNCTION_ARGS)
{
	bool found;

	polar_make_pg_directory("test_slru");
	test_lock = (LWLockMinimallyPadded *)
		ShmemInitStruct("test lock", sizeof(LWLockMinimallyPadded), &found);
	LWLockInitialize(&test_lock->lock, LWTRANCHE_FIRST_USER_DEFINED);

	test_slru_ctl.PagePrecedes = test_slru_page_precedes;
	IsUnderPostmaster = false;
	SimpleLruInit(&test_slru_ctl, "test_slru",
			10, 0, &test_lock->lock, "test_slru", LWTRANCHE_FIRST_USER_DEFINED, false);
	IsUnderPostmaster = true;
	test_slru_page_physical_exists();

	//Test error when promote slru without cache
	test_promote_slru_with_cache_err();
	//Test error when cache write error
	test_promote_slru_with_cache_dir_err();

	PG_RETURN_VOID();
}

PG_FUNCTION_INFO_V1(test_slru_slot_size_config);
/*
 * SQL-callable entry point to perform all tests.
 *
 * If a 1% false positive threshold is not met, emits WARNINGs.
 *
 * See README for details of arguments.
 */
Datum
test_slru_slot_size_config(PG_FUNCTION_ARGS)
{
	polar_enable_shared_storage_mode = true;
	// CLOGShmemBuffers()
	polar_clog_slot_size = Max(4, NBuffers / 512 * 8) - 1;
	Assert(CLOGShmemBuffers() == polar_clog_slot_size);
	
	polar_clog_slot_size = Max(4, NBuffers / 512 * 8) + 1;
	Assert(CLOGShmemBuffers() == Max(4, NBuffers / 512 * 8));

	// CommitTsShmemBuffers()
	polar_committs_buffer_slot_size = Max(4, NBuffers / 1024 * 8) - 1;
	Assert(CommitTsShmemBuffers() == polar_committs_buffer_slot_size);

	polar_committs_buffer_slot_size = Max(4, NBuffers / 1024 * 8) + 1;
	Assert(CommitTsShmemBuffers() == Max(4, NBuffers / 1024 * 8));

	// just check default value
	Assert(polar_mxact_offset_buffer_slot_size == NUM_MXACTOFFSET_BUFFERS);
	Assert(polar_mxact_member_buffer_slot_size == NUM_MXACTMEMBER_BUFFERS);
	Assert(polar_subtrans_buffer_slot_size == NUM_SUBTRANS_BUFFERS);
	Assert(polar_async_buffer_slot_size == NUM_ASYNC_BUFFERS);
	Assert(polar_oldserxid_buffer_slot_size == NUM_OLDSERXID_BUFFERS);
	Assert(polar_async_buffer_slot_size == NUM_MXACTOFFSET_BUFFERS);

	PG_RETURN_VOID();
}

PG_FUNCTION_INFO_V1(test_slru_hash_index);
Datum
test_slru_hash_index(PG_FUNCTION_ARGS)
{
	bool found;
	int slot_num = PG_GETARG_INT32(0);
	int page_num = PG_GETARG_INT32(1);
	int test_num = PG_GETARG_INT32(2);

	test_successor_list(slot_num);

	polar_make_pg_directory("test_slru_hash_index");
	test_lock = (LWLockMinimallyPadded *)
		ShmemInitStruct("test lock hash index", sizeof(LWLockMinimallyPadded), &found);
	LWLockInitialize(&test_lock->lock, LWTRANCHE_FIRST_USER_DEFINED);

	slru_hash_index_ctl.PagePrecedes = test_slru_page_precedes;
	IsUnderPostmaster = false;
	SimpleLruInit(&slru_hash_index_ctl, "test_slru_hash_index",
			slot_num, 0, &test_lock->lock, "test_slru_hash_index", LWTRANCHE_FIRST_USER_DEFINED, false);
	IsUnderPostmaster = true;
	test_slru_hash_index_internal(slot_num, page_num, test_num);

	PG_RETURN_VOID();
}
