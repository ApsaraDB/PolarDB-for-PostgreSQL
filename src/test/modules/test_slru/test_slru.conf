polar_enable_slru_hash_index=on
