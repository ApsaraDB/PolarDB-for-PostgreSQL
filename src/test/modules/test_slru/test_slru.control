comment = 'Test code for slru'
default_version = '1.0'
module_pathname = '$libdir/test_slru'
relocatable = true
