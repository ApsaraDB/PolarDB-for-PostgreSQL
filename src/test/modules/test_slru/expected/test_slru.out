CREATE EXTENSION test_slru;
-- See README for explanation of arguments:
SELECT test_slru();
WARNING:  Failed to do io for local cache, file_type=1, io_permission=F
 test_slru 
-----------
 
(1 row)

SELECT test_slru_slot_size_config();
 test_slru_slot_size_config 
----------------------------
 
(1 row)

SELECT test_slru_hash_index(10, 256, 10000);
 test_slru_hash_index 
----------------------
 
(1 row)

